Use insertion sort instead of sort.Stable (better performance and less GC pressure)

jonasfj opened the pull request:
Basically the `interface{}` casting magic is very slow.
Even worse it makes allocations, the two benchmarks shows before/after
adding the naive insertion sort implementation:

```
Before:
BenchmarkDigestOverNormalDistribution-8   7773315763 ns/op  2573680 B/op   78156 allocs/op
After:
BenchmarkDigestOverNormalDistribution-8   6224821116 ns/op    73680 B/op      31 allocs/op
```

Notice the dramatic difference in number of allocations.
Summary, `sort.Stable` uses a lot of allocations when operation on structs like `Centroid`. Clearly this hurts the GC, etc...

Reproduce tests with `go test -bench . -benchmem -memprofilerate 1`. You might have to adjust the number of iterations to get enough for the allocations to register.

jonasfj committed:
Improve performance - avoid using sort.Stable

Basically the interface{} casting magic is very slow.
This makes sort.Stable rather slow. In fact the sortByMean gets
heap allocated, as it is passed by reference.
The two benchmarks shows before/after adding the naive insertion
sort implementation:

Before:
BenchmarkDigestOverNormalDistribution-8
                    7773315763 ns/op   2573680 B/op   78156 allocs/op
After:
BenchmarkDigestOverNormalDistribution-8
                    6224821116 ns/op     73680 B/op      31 allocs/op

The benchmark above is made by forcing 10M numbers in a single benchmark
iteration to avoid the allocations being divided by number of values
added to the t-digest.

jonasfj commented:
Note: It took a lot of work to track this down in my applications...
I could just see that something was allocating memory in a tight loop, finding out that it was sort was not easy... The `append()` use here is very careful to reuse buffers, and it took a while to eliminate that :)

---

I sort of suspect it would be smarter to avoid using `append()` and manually ensure the buffers are doubled in size... As the current performance sort of relies on a golang implementation detail. Namely that it can detect that the discarded part of the buffer isn't used... Technically, I doubt anything promises that `mySlice = mySlice[:0]` will remember that `mySlice[0:]` isn't referenced. After all, `append` does return a new slice/array...

But that's for another day, go clearly does the thing here.

snaury commented:
Relative centroid position is important for interpolation, and using unstable sort breaks precision guarantee of `4q(1-q)`. It's easiest to show an edge case: imagine you have a million of exactly the same values (`maj_value`) and you add them to tdigest first, which when merged into centroids are at most `4Nq(1-q)/C` in size (where `N=1000000` and `C=compression`), so their weight is very small at edges, and 1% of total weight towards the center (with `compression=100`). When you add two new values, one below the majority (`min_value`) and one above the majority (`max_value`), your centroids would be resorted and in unstable sort the order is unreliable. Let's look at the worst case, where centroid with 1% of the total weight (which is 100000) becomes right next to the new `min_value`. If we didn't use interpolation that wouldn't be a problem, however in our case for `q` from `0` to `0.005` there would be interpolation between `min_value` and `maj_value`. In the middle of that range, at `q=0.0025`, the maximum allowed error is around `99.75` values (if we give any result that has bigger distance than that from the true result we are violating precision guarantees). Our result would be half-way between `min_value` and `maj_value`, but we should remember that originally we're in the middle of a sea of 50000 exactly the same values and the only possible result is exactly `maj_value`, not half-way between `min_value` and `maj_value`.

That is why using unstable sort, while tempting, is incorrect.

jonasfj commented:
@snaury, this features an **insertion sort**, it is a **stable sort**, see:
https://en.wikipedia.org/wiki/Insertion_sort

The test case included in the PR also validates that the sort is indeed stable.

jonasfj commented:
@snaury, it's faster because go built-in stable sort relies on `interface{}`... So it's casting things around, doing allocations and all sorts of things a sorting function shouldn't be doing :)

This is really an excellent of case "wow, go needs generics". It makes me really sad that it's so bad at sorting.

snaury commented:
@jonasfj ah, sorry! I have looked at the title before I looked in the code.

jonasfj commented:
@snaury, fair enough... it's also very weird that it's the builtin sort that's being slow and allocating memory. I was very surprised, engineers tend to love showing off their sort implementations :)

Also I didn't realize the stable part was so important...
I guess it really matters if you use the `MergingDigest.Merge(Centroid)` function to merged two digests.

The common case is people using `MergingDigest.Add(float64, 1)` with weight one, to add new observations. And those centroids could be sorted with an unstable sort. However, I doubt it worth it to make a boolean to track if `MergingDigest.Merge` has been called and use quick-sort if it haven't.
But that could be another optimization to play with one day.

snaury commented:
> it's also very weird that it's the builtin sort that's being slow and allocating memory

Funny enough it actually doesn't! :) You can achieve the same effect by caching the allocated `sortByMean` structure:

``` diff
diff --git a/tdigest.go b/tdigest.go
index e3845cf..522d372 100644
--- a/tdigest.go
+++ b/tdigest.go
@@ -19,6 +19,7 @@ type MergingDigest struct {
    count       int64
    compression float64
    MaxUnmerged int
+   sorter      sortByMean
 }

 // New creates a new MergingDigest with the given compression
@@ -79,7 +80,8 @@ func (digest *MergingDigest) Compress() {
    if len(digest.unmerged) == 0 {
        return
    }
-   sort.Stable(sortByMean(digest.unmerged))
+   digest.sorter = digest.unmerged
+   sort.Stable(&digest.sorter)
    sum := int64(0)
    m := 0
    i := 0
```

jonasfj commented:
Oh, that's interesting... so it was the `sortByMean` structure, I would totally have assumed that thing was stack allocated :)
But I guess the go compiler can't see that `sort.Stable` doesn't keep a reference to it...
Hmm... a bit confusing since it's passed as value, but I guess `interface{}` results in boxing.

So yeah, we'll need to keep the sorter around. Or reimplement sorting, I'll try to play with it this weekend maybe..

snaury commented:
Nevertheless, while allocations are not actually important (effect on performance is negligible, as expected), getting rid of indirect method calls seems like what actually helps. But the current test doesn't use `b.N`, so can't merge as is, will look at it later.

snaury commented:
> Hmm... a bit confusing since it's passed as value, but I guess `interface{}` results in boxing.

Passing by value is not important, since it is converted to the interface first, and yes the actual value ends up on the heap as a result. The problem seems to be with escape analysis, but needs more research.

jonasfj commented:
`b.N` nice, I didn't see that !!! :) thanks... (I'm a bit new at go -- but it's so lovely simple)

Indirect method calls helps on perf sure... From playing with my app (stats aggregator) it seems that the small allocations adds up to more memory usage, at-least when under heavy pressure.

Thanks, for the rapid comments, I'll play a bit more with this...

jonasfj commented:
Added a comment to ensure that we remember it must be stable sort..

And changed benchmark to also used `b.N`, I do increments of 10k values to make allocations observable.. If allocations are divided by 10M as the number of operations done, it won't show up...
Probably because you got the amortized complexity right :)

snaury commented:
Ok, I'm still wary of merging this pull request, because using pure insertion sort means it may hurt in some worst cases, for example, these are benchmarks before using insertionSort:

```
BenchmarkDigest-4                        5000000           293 ns/op
BenchmarkDigest10BatchesNormal-4            2000        814849 ns/op
BenchmarkDigest10BatchesSequential-4        5000        304811 ns/op
BenchmarkDigest10BatchesReverse-4           2000        554384 ns/op
```

And these are benchmarks after using insertionSort:

```
BenchmarkDigest-4                       10000000           239 ns/op
BenchmarkDigest10BatchesNormal-4            2000        676257 ns/op
BenchmarkDigest10BatchesSequential-4       10000        310185 ns/op
BenchmarkDigest10BatchesReverse-4           2000        917148 ns/op
```

As it can be seen when adding numbers in reverse it can be quite a bit slower. What's more, faster sorting means we could probably raise the default for MaxUnmerged higher, however since it's insertion sort the more elements there are the worst might be the result.

I'm going to try to just take `sort.go` from go sources and specialise it for centroids.

snaury commented:
And using a specialised version of `sort.go` I get this:

```
BenchmarkDigest-4                       10000000           226 ns/op
BenchmarkDigest10BatchesNormal-4            2000        621008 ns/op
BenchmarkDigest10BatchesSequential-4       10000        303699 ns/op
BenchmarkDigest10BatchesReverse-4           5000        457346 ns/op
```

And raising DefaultMaxUnmerged a little gives even better performance. Please see #2.

jonasfj commented:
I agree, insertionsort is not ideal... :)
I like #2, that seems to solve everything.

Note: maybe we should make it easy to set MaxUnmerged in the constructor. Use make(..,.., MaxUnmerged) to get a initial capacity right.
Disadvantage of higher MaxUnmerged is higher memory usage which matters if you have a lot of tdigest...
I suspect we could also use sync.Pool for the merged array as its only used temporarily (from a quick look at it).

By the way, it sad to see gen not offering to specialize sort.go:
https://github.com/clipperhouse/gen
That would have been ideal, but from their docs it seems they duplicate the Slice...
